File: crates/kiln-train/src/pi_trajectory.rs
//! Pi session JSONL normalization into kiln's canonical trajectory schema.

use std::path::Path;

use anyhow::{Context, Result};
use serde_json::Value;

use crate::ChatMessage;
use crate::trajectory::{ScoredRollout, TurnKind, TurnSegment};

#[derive(Debug, Clone, Default)]
pub struct PiSessionParse {
    pub prompt_messages: Vec<ChatMessage>,
    pub trajectory: Vec<TurnSegment>,
    pub warnings: Vec<String>,
}

pub fn parse_pi_session_jsonl(path: &Path, include_context: bool) -> Result<PiSessionParse> {
    let raw = std::fs::read_to_string(path).with_context(|| format!("read {}", path.display()))?;
    Ok(parse_pi_session_str(&raw, include_context))
}

pub fn parse_pi_session_str(raw: &str, include_context: bool) -> PiSessionParse {
    let mut values = Vec::new();
    let mut warnings = Vec::new();
    for (line_idx, line) in raw.lines().enumerate() {
        let trimmed = line.trim();
        if trimmed.is_empty() {
            continue;
        }
        match serde_json::from_str::<Value>(trimmed) {
            Ok(value) => values.push(value),
            Err(err) => warnings.push(format!(
                "skipped malformed JSONL line {}: {err}",
                line_idx + 1
            )),
        }
    }
    parse_pi_session_values(&values, include_context, warnings)
}

pub fn parse_pi_session_values(
    values: &[Value],
    include_context: bool,
    mut warnings: Vec<String>,
) -> PiSessionParse {
    let mut prompt_messages = Vec::new();
    let mut trajectory = Vec::new();
    for value in values {
        if !is_pi_message_event(value) {
            continue;
        }
        let Some(message) = value.get("message").and_then(Value::as_object) else {
            warnings.push("skipped Pi message event with non-object message".to_string());
            continue;
        };
        let role = message.get("role").and_then(Value::as_str).unwrap_or("");
        let content = message.get("content").unwrap_or(&Value::Null);
        match role {
            "system" | "user" => {
                let text = stringify_content(content);
                if text.is_empty() {
                    continue;
                }
                if include_context {
                    trajectory.push(TurnSegment {
                        role: role.to_string(),
                        content: text,
                        kind: TurnKind::Context,
                        tool_call_id: None,
                        warning_prefix_len: None,
                    });
                } else {
                    prompt_messages.push(ChatMessage {
                        role: role.to_string(),
                        content: text,
                    });
                }
            }
            "assistant" => {
                let rendered = render_assistant_content(content, &mut warnings);
                if !rendered.is_empty() {
                    trajectory.push(TurnSegment {
                        role: "assistant".to_string(),
                        content: rendered,
                        kind: TurnKind::Action,
                        tool_call_id: None,
                        warning_prefix_len: None,
                    });
                }
            }
            "tool" | "toolResult" => {
                let (rendered, tool_call_id) = render_tool_result_content(content, &mut warnings);
                if !rendered.is_empty() {
                    trajectory.push(TurnSegment {
                        role: "tool".to_string(),
                        warning_prefix_len: detect_warning_prefix(&rendered),
                        content: rendered,
                        kind: TurnKind::Observation,
                        tool_call_id,
                    });
                }
            }
            other => warnings.push(format!("skipped Pi message with unknown role {other:?}")),
        }
    }
    PiSessionParse {
        prompt_messages,
        trajectory,
        warnings,
    }
}

pub fn is_pi_message_event(value: &Value) -> bool {
    value.get("type").and_then(Value::as_str) == Some("message") && value.get("message").is_some()
}

pub fn flatten_action_text(trajectory: &[TurnSegment]) -> String {
    trajectory
        .iter()
        .filter(|segment| segment.kind == TurnKind::Action)
        .map(|segment| segment.content.as_str())
        .collect::<Vec<_>>()
        .join("<TURN_BREAK>")
}

pub fn scored_rollout_from_pi_session(
    raw: &str,
    reward: f64,
    include_context: bool,
) -> ScoredRollout {
    let parsed = parse_pi_session_str(raw, include_context);
    let text = {
        let flattened = flatten_action_text(&parsed.trajectory);
        if flattened.is_empty() {
            "(empty)".to_string()
        } else {
            flattened
        }
    };
    ScoredRollout {
        text,
        reward,
        trajectory: parsed.trajectory,
    }
}

pub fn stringify_content(content: &Value) -> String {
    match content {
        Value::String(text) => text.clone(),
        Value::Array(blocks) => blocks
            .iter()
            .filter_map(|block| {
                if let Some(text) = block.as_str() {
                    Some(text.to_string())
                } else {
                    block
                        .get("text")
                        .and_then(Value::as_str)
                        .map(ToString::to_string)
                }
            })
            .collect::<Vec<_>>()
            .join(""),
        Value::Null => String::new(),
        other => other.to_string(),
    }
}

fn render_assistant_content(content: &Value, warnings: &mut Vec<String>) -> String {
    let Some(blocks) = content.as_array() else {
        return stringify_content(content);
    };
    let mut parts = Vec::new();
    for block in blocks {
        let Some(kind) = block.get("type").and_then(Value::as_str) else {
            continue;
        };
        match kind {
            "text" => {
                if let Some(text) = block.get("text").and_then(Value::as_str) {
                    parts.push(text.to_string());
                }
            }
            "thinking" => {
                if let Some(text) = block.get("thinking").and_then(Value::as_str) {
                    parts.push(format!("<think>{text}</think>"));
                }
            }
            "toolCall" => {
                let name = block.get("name").and_then(Value::as_str).unwrap_or("");
                let name_json = serde_json::to_string(name).unwrap_or_else(|_| "\"\"".to_string());
                let args = block.get("input").unwrap_or(&Value::Null);
                let args_json = if args.is_null() {
                    "{}".to_string()
                } else {
                    serde_json::to_string(args).unwrap_or_else(|_| "{}".to_string())
                };
                parts.push(format!(
                    "<tool_call>{{\"name\": {name_json}, \"arguments\": {args_json}}}</tool_call>"
                ));
            }
            other => warnings.push(format!(
                "skipped unsupported assistant block type {other:?}"
            )),
        }
    }
    parts.join("")
}

fn render_tool_result_content(
    content: &Value,
    warnings: &mut Vec<String>,
) -> (String, Option<String>) {
    let Some(blocks) = content.as_array() else {
        return (stringify_content(content), None);
    };
    let mut parts = Vec::new();
    let mut tool_call_id = None;
    for block in blocks {
        let Some(kind) = block.get("type").and_then(Value::as_str) else {
            continue;
        };
        match kind {
            "toolResult" => {
                let result_content = block.get("content").unwrap_or(&Value::Null);
                parts.push(stringify_content(result_content));
                if tool_call_id.is_none() {
                    tool_call_id = block
                        .get("toolCallId")
                        .or_else(|| block.get("tool_call_id"))
                        .or_else(|| block.get("id"))
                        .and_then(Value::as_str)
                        .map(ToString::to_string);
                }
            }
            "text" => {
                if let Some(text) = block.get("text").and_then(Value::as_str) {
                    parts.push(text.to_string());
                }
            }
            other => warnings.push(format!("skipped unsupported tool block type {other:?}")),
        }
    }
    (parts.join(""), tool_call_id)
}

pub fn detect_warning_prefix(text: &str) -> Option<usize> {
    if !text.starts_with("WARNINGS:\n") {
        return None;
    }
    if let Some(idx) = text.find("<command_output>") {
        if idx > 0 {
            return Some(idx);
        }
    }
    if let Some(idx) = text.find("\n\n") {
        if idx > 0 {
            return Some(idx + 2);
        }
    }
    Some("WARNINGS:\n".len())
}

#[cfg(test)]
mod tests {
    use super::*;

    fn jsonl(lines: &[Value]) -> String {
        let mut out = String::new();
        for line in lines {
            out.push_str(&serde_json::to_string(line).unwrap());
            out.push('\n');
        }
        out
    }

    #[test]
    fn parses_pi_0751_session_like_python_parser() {
        let raw = jsonl(&[
            serde_json::json!({"type":"message","message":{"role":"system","content":[{"type":"text","text":"Python assistant"}]}}),
            serde_json::json!({"type":"message","message":{"role":"user","content":[{"type":"text","text":"Print 42"}]}}),
            serde_json::json!({"type":"message","message":{"role":"assistant","content":[{"type":"thinking","thinking":"I should use bash"},{"type":"toolCall","name":"bash","input":{"cmd":"python3 -c 'print(42)'"},"id":"c1"}]}}),
            serde_json::json!({"type":"message","message":{"role":"tool","content":[{"type":"toolResult","content":"42\n","toolCallId":"c1"}]}}),
            serde_json::json!({"type":"message","message":{"role":"assistant","content":[{"type":"text","text":"Done - the program printed 42."}]}}),
        ]);

        let parsed = parse_pi_session_str(&raw, false);

        assert_eq!(parsed.prompt_messages.len(), 2);
        assert_eq!(parsed.trajectory.len(), 3);
        assert_eq!(
            parsed.trajectory.iter().map(|s| s.kind).collect::<Vec<_>>(),
            vec![TurnKind::Action, TurnKind::Observation, TurnKind::Action]
        );
        assert!(parsed.trajectory[0].content.contains("<think>I should use bash</think>"));
        assert!(parsed.trajectory[0].content.contains("<tool_call>"));
        assert!(parsed.trajectory[0].content.contains("\"arguments\""));
        assert_eq!(parsed.trajectory[1].role, "tool");
        assert_eq!(parsed.trajectory[1].content, "42\n");
        assert_eq!(parsed.trajectory[1].tool_call_id.as_deref(), Some("c1"));
        assert_eq!(parsed.trajectory[2].content, "Done - the program printed 42.");
    }

    #[test]
    fn parses_pi_0753_tool_result_role_like_python_parser() {
        let raw = jsonl(&[
            serde_json::json!({"type":"message","message":{"role":"assistant","content":[{"type":"toolCall","name":"bash","input":{"cmd":"echo ok"},"id":"c1"}]}}),
            serde_json::json!({"type":"message","message":{"role":"toolResult","content":[{"type":"text","text":"ok\n"}]}}),
        ]);

        let parsed = parse_pi_session_str(&raw, false);

        assert_eq!(parsed.trajectory.len(), 2);
        assert_eq!(parsed.trajectory[1].role, "tool");
        assert_eq!(parsed.trajectory[1].kind, TurnKind::Observation);
        assert_eq!(parsed.trajectory[1].content, "ok\n");
    }

    #[test]
    fn include_context_emits_system_and_user_segments() {
        let raw = jsonl(&[
            serde_json::json!({"type":"message","message":{"role":"system","content":[{"type":"text","text":"sys"}]}}),
            serde_json::json!({"type":"message","message":{"role":"user","content":[{"type":"text","text":"do it"}]}}),
            serde_json::json!({"type":"message","message":{"role":"assistant","content":[{"type":"text","text":"ok"}]}}),
        ]);

        let parsed = parse_pi_session_str(&raw, true);

        assert!(parsed.prompt_messages.is_empty());
        assert_eq!(parsed.trajectory.len(), 3);
        assert_eq!(
            parsed.trajectory.iter().map(|s| s.kind).collect::<Vec<_>>(),
            vec![TurnKind::Context, TurnKind::Context, TurnKind::Action]
        );
    }

    #[test]
    fn warning_prefix_and_flattening_match_python_parser_contract() {
        let warning = "WARNINGS:\n- something bad\n\nactual output";
        let raw = jsonl(&[
            serde_json::json!({"type":"message","message":{"role":"assistant","content":[{"type":"text","text":"first"}]}}),
            serde_json::json!({"type":"message","message":{"role":"tool","content":[{"type":"toolResult","content":warning}]}}),
            serde_json::json!({"type":"message","message":{"role":"assistant","content":[{"type":"text","text":"second"}]}}),
        ]);

        let parsed = parse_pi_session_str(&raw, false);

        let warning_prefix_len = parsed.trajectory[1].warning_prefix_len.unwrap();
        assert!(warning_prefix_len > 0);
        assert!(warning_prefix_len <= warning.find("actual output").unwrap());
        assert_eq!(flatten_action_text(&parsed.trajectory), "first<TURN_BREAK>second");
        let rollout = scored_rollout_from_pi_session(&raw, 1.0, false);
        assert_eq!(rollout.text, "first<TURN_BREAK>second");
        assert_eq!(rollout.trajectory.len(), parsed.trajectory.len());
        assert_eq!(rollout.trajectory[0].content, parsed.trajectory[0].content);
        assert_eq!(rollout.trajectory[1].warning_prefix_len, Some(warning_prefix_len));
    }
}

File: capabilities/lib/pi_trajectory.py
"""Parse pi session JSONL into the canonical kiln Trajectory schema.

The canonical schema (kiln-train::trajectory::ScoredRollout) carries an
optional list of TurnSegments where each segment has:

    {
        "role": "system" | "user" | "assistant" | "tool",
        "content": <raw text the model emitted or saw>,
        "kind": "context" | "action" | "observation",
        "tool_call_id": <optional correlation id>,
        "warning_prefix_len": <bytes of harness warning to strip from env mask>
    }

Action segments are policy-gradient targets (assistant turns).
Observation segments are ECHO's env-CE targets (tool results / environment).
Context segments are system/user prompts — no gradient.

This module reads pi's `~/.pi/sessions/<uuid>.jsonl` format. The format
(verified against capabilities/caps/pi-doctest/rollout.py on
2026-05-18):

    {"type": "message", "message": {
        "role": "system" | "user" | "assistant" | "tool",
        "content": [{
            "type": "text" | "thinking" | "toolCall" | "toolResult",
            ...role-specific fields...
        }]
    }}

Tool-call args have appeared under both b["input"] and b["arguments"]
across pi versions. Tool result blocks use b["content"] (string) and
b["toolCallId"] for correlation.

See `docs/plans/echo-integration-plan.md` §3.3 and §B.8 for the design.
"""
from __future__ import annotations

import json
from pathlib import Path
from typing import Any, Iterator, Optional


# Renderers from pi's content blocks to the Qwen-XML form the chat template
# expects. Mirror what rollout.py:280-318 was doing inline; centralised here
# so every cap (not just pi-doctest) gets the same conversion.
def _render_assistant_block(b: dict) -> Optional[str]:
    bt = b.get("type")
    if bt == "text" and isinstance(b.get("text"), str):
        return b["text"]
    if bt == "thinking" and isinstance(b.get("thinking"), str):
        # Qwen-style <think> tags so the chat template round-trip is closer
        # to what the model emitted at sample time.
        return f"<think>{b['thinking']}</think>"
    if bt == "toolCall":
        name = b.get("name", "")
        # Pi version drift: 0.75.1 used "input"; current builds emit
        # "arguments". Preserve both so action tokens retain tool payloads.
        args_obj = b.get("input") or b.get("arguments") or {}
        args_json = json.dumps(args_obj)
        return (
            f"<tool_call>"
            f'{{"name": "{name}", "arguments": {args_json}}}'
            f"</tool_call>"
        )
    return None


def _render_assistant_content(blocks: list) -> str:
    parts: list[str] = []
    for b in blocks:
        if not isinstance(b, dict):
            continue
        rendered = _render_assistant_block(b)
        if rendered is not None:
            parts.append(rendered)
    return "".join(parts)


def _extract_tool_call_ids(blocks: list) -> list[str]:
    """Tool-call IDs emitted by the assistant in this turn; matched against
    subsequent tool-result events for correlation."""
    ids: list[str] = []
    for b in blocks:
        if not isinstance(b, dict):
            continue
        if b.get("type") == "toolCall":
            tid = b.get("id") or b.get("toolCallId")
            if isinstance(tid, str):
                ids.append(tid)
    return ids


def _render_tool_result_content(blocks: list) -> tuple[str, Optional[str]]:
    """Render a tool-role message's content blocks. Returns
    (rendered_text, tool_call_id-or-None).

    Handles two pi formats:
      - pi 0.75.1: role="tool", content=[{type:"toolResult", content:..., toolCallId:...}]
      - pi 0.75.3: role="toolResult", content=[{type:"text", text:...}]
    """
    parts: list[str] = []
    tool_call_id: Optional[str] = None
    for b in blocks:
        if not isinstance(b, dict):
            continue
        bt = b.get("type")
        if bt == "toolResult":
            content = b.get("content", "")
            if isinstance(content, str):
                parts.append(content)
            elif isinstance(content, list):
                for inner in content:
                    if isinstance(inner, dict) and isinstance(inner.get("text"), str):
                        parts.append(inner["text"])
            tid = b.get("toolCallId") or b.get("tool_call_id") or b.get("id")
            if isinstance(tid, str) and tool_call_id is None:
                tool_call_id = tid
        elif bt == "text" and isinstance(b.get("text"), str):
            parts.append(b["text"])
    return "".join(parts), tool_call_id


def _detect_warning_prefix(text: str) -> Optional[int]:
    """Detect the harness warning prefix kiln emits when a tool call fails
    parsing or violates format constraints. Returns the byte length of the
    prefix (so the masker can advance past it), or None if no warning.

    Paper §3.2: warning tokens memorize within ~60 steps and stop providing
    useful gradient; ECHO's env_mask excludes them when
    MaskConfig::warning_filter is true (the default).
    """
    if not text.startswith("WARNINGS:\n"):
        return None
    # Find the start of real terminal output. kiln's harness wraps it in
    # <command_output>...</command_output> — strip up to that.
    idx = text.find("<command_output>")
    if idx > 0:
        return idx
    # Fallback: any double-newline after the warning block.
    idx = text.find("\n\n")
    if idx > 0:
        return idx + 2
    return len("WARNINGS:\n")


def _iter_events(path: Path) -> Iterator[dict]:
    """Yield JSON-decoded events from a pi session JSONL, silently skipping
    malformed lines. Mirrors the lenient parse in rollout.py."""
    if not path or not path.exists():
        return
    with path.open() as f:
        for line in f:
            line = line.strip()
            if not line:
                continue
            try:
                yield json.loads(line)
            except (json.JSONDecodeError, ValueError):
                continue


def parse_pi_session(
    session_path: Path,
    include_context: bool = False,
) -> list[dict]:
    """Parse a pi session JSONL file into a list of TurnSegment dicts.

    Returns a list compatible with the canonical kiln Trajectory schema:

        [
            {"role": "system",    "content": "...", "kind": "context"},
            {"role": "user",      "content": "...", "kind": "context"},
            {"role": "assistant", "content": "...", "kind": "action"},
            {"role": "tool",      "content": "...", "kind": "observation",
             "tool_call_id": "...", "warning_prefix_len": null|int},
            ...
        ]

    When `include_context=False` (default) system/user turns are filtered out
    because they belong in the surrounding `AgenticGroup.messages` field
    rather than per-rollout trajectory. When `include_context=True` every
    turn is emitted, useful for end-to-end fidelity.
    """
    segments: list[dict] = []

    for event in _iter_events(session_path):
        if event.get("type") != "message":
            continue
        msg = event.get("message") or {}
        role = msg.get("role")
        content = msg.get("content")

        if role in ("system", "user"):
            if not include_context:
                continue
            text = _stringify_content(content)
            if text:
                segments.append({"role": role, "content": text, "kind": "context"})

        elif role == "assistant":
            if isinstance(content, list):
                rendered = _render_assistant_content(content)
                if rendered:
                    segments.append(
                        {"role": role, "content": rendered, "kind": "action"}
                    )

        elif role in ("tool", "toolResult"):
            if isinstance(content, list):
                rendered, tool_call_id = _render_tool_result_content(content)
                if rendered:
                    # Normalize the role label to "tool" — the Qwen chat
                    # template only accepts that name; pi 0.75.1 used
                    # "tool" while pi 0.75.3 emits "toolResult" but both
                    # carry the same observation tokens. (chat template
                    # raised "Unexpected message role" otherwise.)
                    seg: dict[str, Any] = {
                        "role": "tool",
                        "content": rendered,
                        "kind": "observation",
                    }
                    if tool_call_id is not None:
                        seg["tool_call_id"] = tool_call_id
                    warning_prefix_len = _detect_warning_prefix(rendered)
                    if warning_prefix_len is not None:
                        seg["warning_prefix_len"] = warning_prefix_len
                    segments.append(seg)

    return segments


def _stringify_content(content: Any) -> str:
    """Flatten a pi content blob into a single string (used for system/user
    blocks which are typically already strings, but handle list form too)."""
    if isinstance(content, str):
        return content
    if isinstance(content, list):
        parts: list[str] = []
        for b in content:
            if isinstance(b, dict) and isinstance(b.get("text"), str):
                parts.append(b["text"])
            elif isinstance(b, str):
                parts.append(b)
        return "".join(parts)
    return str(content) if content is not None else ""


def flatten_action_text(trajectory: list[dict]) -> str:
    """Mirror of `ScoredRollout::from_trajectory`'s text-flattening logic in
    Rust: join Action segment contents with `<TURN_BREAK>` so the legacy
    single-string `text` field stays populated even when `trajectory` is
    the canonical representation."""
    return "<TURN_BREAK>".join(
        seg["content"] for seg in trajectory if seg.get("kind") == "action"
    )


def build_scored_rollout(
    session_path: Path,
    reward: float,
    include_context: bool = False,
) -> dict:
    """One-shot helper: parse a pi session and emit a ScoredRollout-shaped
    dict. The result deserializes into kiln-train's ScoredRollout type
    via serde with no further transformation.

        {
            "text": "<flattened action segments joined by <TURN_BREAK>>",
            "reward": <reward>,
            "trajectory": [<TurnSegment>, ...],
        }
    """
    trajectory = parse_pi_session(session_path, include_context=include_context)
    text = flatten_action_text(trajectory) or "(empty)"
    return {"text": text, "reward": reward, "trajectory": trajectory}
